bag_of_words/code/classifiers.py

```diff
@@ -0,0 +1,89 @@
+import pickle, heapq
+from itertools import compress
+import numpy as np
+from sklearn.model_selection import cross_validate
+from sklearn.svm import LinearSVC
+from sklearn.tree import DecisionTreeClassifier
+from sklearn.ensemble import RandomForestClassifier
+
+indir = "../data/processed/"
+
+with open(indir + "bag_of_words.pkl", 'rb') as fin:
+    bow = pickle.load(fin)  # doc-by-word count sparse matrix
+    bow_rownames = pickle.load(fin)  # list of PMIDs
+    bow_colnames = pickle.load(fin)  # list of distinct words
+
+with open(indir + "doc_labels.pkl", 'rb') as fin:
+    labels = pickle.load(fin)  # doc labels
+
+# Remove docs that have no labels
+labels_pos_ind = (labels.sum(axis=1) > 0)
+labels_pos_ind.sum() * 1.0 / len(labels_pos_ind)
+bow = bow[labels_pos_ind.values,:]
+bow_rownames = list(compress(bow_rownames, labels_pos_ind.values))
+labels = labels[labels_pos_ind.values]
+
+# Take the topk tokens
+topk = 1000
+df = (bow > 0).sum(axis=0)
+bow = bow[:,(df > 0).A1]  # Remove words with zero occurance in the sub-corpus
+df = (bow > 0).sum(axis=0)
+tf = bow.sum(axis=0)
+idf = np.log( (len(bow_rownames) + 1) / df )
+tfidf = np.multiply(tf, idf)

```

Comment:
I recommend that we use the TF-IDF vectorizers, which are optimized, and also will ensure we don't make any simple mistakes with these calculations.

Reply:
Those vectorizers also come with built-in methods for selecting cutoffs.